subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/ImplementationSnapshot.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot;
+
+import org.gradle.internal.hash.ClassLoaderHierarchyHasher;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.snapshot.impl.DefaultImplementationSnapshot;
+import org.gradle.internal.snapshot.impl.LambdaImplementationSnapshot;
+import org.gradle.internal.snapshot.impl.UnknownClassloaderImplementationSnapshot;
+
+import javax.annotation.Nullable;
+
+/**
+ * Identifies a type in a classloader hierarchy. The type is identified by its name,
+ * the classloader hierarchy by its hash code.
+ */
+public abstract class ImplementationSnapshot implements ValueSnapshot {
+    private static final String GENERATED_LAMBDA_CLASS_SUFFIX = "$$Lambda$";
+
+    private final String typeName;
+
+    public static ImplementationSnapshot of(Class<?> type, ClassLoaderHierarchyHasher classLoaderHasher) {
+        String className = type.getName();
+        return of(className, classLoaderHasher.getClassLoaderHash(type.getClassLoader()), type.isSynthetic() && isLambdaClassName(className));
+    }
+
+    public static ImplementationSnapshot of(String className, @Nullable HashCode classLoaderHash) {
+        return of(className, classLoaderHash, isLambdaClassName(className));
+    }
+
+    private static ImplementationSnapshot of(String typeName, @Nullable HashCode classLoaderHash, boolean lambda) {
+        if (classLoaderHash == null) {
+            return new UnknownClassloaderImplementationSnapshot(typeName);
```

Comment:
Isn't this a package cycle? `snapshot` -> `impl` -> `snapshot`?

Reply:
Good point, I moved `ImplementationSnapshot` back to the `impl` package.